tests/test_sag.py

```diff
@@ -0,0 +1,264 @@
+        self.check_object_exist(self.asic_db,
+            "ASIC_STATE:SAI_OBJECT_TYPE_ROUTER_INTERFACE",
+            {
+                "SAI_ROUTER_INTERFACE_ATTR_TYPE": "SAI_ROUTER_INTERFACE_TYPE_VLAN",
+                "SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS": mac,
+                "SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID": vrf_oid,
+                "SAI_ROUTER_INTERFACE_ATTR_VLAN_ID": vlan_oid
+            }
+        )
+
+    def check_asic_db_route_entry(self, destination, vrf_oid, exist):
+        route_entries = self.asic_db.get_keys("ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY")
+        dest_vrf = [(json.loads(route_entry)["dest"], json.loads(route_entry)["vr"])
+                        for route_entry in route_entries]
+
+        if exist:
+            assert (destination, vrf_oid) in dest_vrf
+        else:
+            assert (destination, vrf_oid) not in dest_vrf
+
+
+    def test_SagAddRemove(self, dvs):
+        self.setup_db(dvs)
+
+        default_mac = "00:00:00:00:00:00"
+        default_vrf_oid = self.get_asic_db_default_vrf_oid()
+        system_mac = self.get_system_mac(dvs)
+
+        interface = "Ethernet0"
+        vlan = "100"
+        vlan_intf = "Vlan{}".format(vlan)
+        self.setup_interface(dvs, interface, vlan)
+
+        ip = "1.1.1.1/24"
+        dvs.add_ip_address(vlan_intf, ip)
+
+        # add SAG global MAC address
+        mac = "00:11:22:33:44:55"
+        self.add_sag_mac(mac)
+        fvs = dvs.get_app_db().wait_for_entry(swsscommon.APP_SAG_TABLE_NAME, "GLOBAL")
+        self.check_app_db_sag_mac(fvs, mac)
+
+        ipv6_ll_route = self.generate_ipv6_link_local_addr(mac, 128)
+        self.check_asic_db_route_entry(str(ipv6_ll_route), default_vrf_oid, exist=True)
+
+        # enable SAG on the VLAN interface
+        self.enable_sag(vlan)
+        fvs = self.app_db.wait_for_field_match(
+            swsscommon.APP_INTF_TABLE_NAME,
+            vlan_intf,
+            {"mac_addr": mac})
+
+        self.check_app_db_intf(fvs, mac, "true")
+        self.check_kernel_intf_mac(dvs, vlan_intf, mac)
+
+        ipv6_ll = self.generate_ipv6_link_local_addr(mac, 64)
+        self.check_kernel_intf_ipv6_addr(dvs, vlan_intf, str(ipv6_ll))
+        self.check_asic_db_router_interface(dvs.getVlanOid(vlan), mac, default_vrf_oid)
+
+        # disable SAG on the VLAN interface
+        self.disable_sag(vlan)
+        fvs = self.app_db.wait_for_field_match(
+            swsscommon.APP_INTF_TABLE_NAME,
+            vlan_intf,
+            {"NULL": "NULL"}
+        )
+
+        self.check_app_db_intf(fvs, default_mac, "false")
+        self.check_kernel_intf_mac(dvs, vlan_intf, system_mac)
+
+        ipv6_ll = self.generate_ipv6_link_local_addr(system_mac, 64)
+        self.check_kernel_intf_ipv6_addr(dvs, vlan_intf, str(ipv6_ll))
+        self.check_asic_db_router_interface(dvs.getVlanOid(vlan), system_mac, default_vrf_oid)
+
+        # delete SAG global MAC address
```

Comment:
Could you please add a case that "delete SAG global MAC address while there is a vlan which enabled SAG"?

Reply:
Sure, I'll add this case and update then.